### PR TITLE
Finagle instrumentation Draft

### PR DESCRIPTION
**Pull request by randomanderson**
Draft pr for finagle. Not ready for review.
Not finished
- Muzzle and version ranges still need
- More tests
- some cleanup

### CODE REVIEW

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java [hunk 1]
```diff
@@ -44,12 +44,22 @@ public static boolean shouldAttachStateToTask(final Object task, final Executor
...
+    if (!task.getClass()
+        .getName()
+        .startsWith("io.netty.util.concurrent.SingleThreadEventExecutor$")) {
+
+      final TraceScope.Continuation continuation = scope.capture();
+      if (state.setContinuation(continuation)) {
+        log.debug("created continuation {} from scope {}, state: {}", continuation, scope, state);
+      } else {
+        continuation.close(false);
+      }
```
Comment (devinsba) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: Oh nice. Based on the comment this seems like the issue I've been seeing with `Akka` I may try this there too

File: dd-java-agent/instrumentation/finagle/src/test/groovy/FinagleServiceInstrumentationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+import datadog.trace.agent.test.AgentTestRunner
+import datadog.trace.agent.test.asserts.TraceAssert
+import datadog.trace.agent.test.utils.PortUtils
+import datadog.trace.api.DDSpanTypes
+import datadog.trace.api.Trace
+import datadog.trace.instrumentation.api.Tags
+import spock.lang.Ignore
+import spock.lang.Shared
+
+class FinagleServiceInstrumentationTest extends AgentTestRunner {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/finagle/src/test/groovy/FinagleServiceInstrumentationTest.groovy]: Can this test extend HttpServerTest (see the spring test for how to do it with handlers)?

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java [hunk 3]
```diff
@@ -44,12 +44,22 @@ public static boolean shouldAttachStateToTask(final Object task, final Executor
...
-      log.debug("created continuation {} from scope {}, state: {}", continuation, scope, state);
-    } else {
-      continuation.close(false);
+
+    // Don't instrument the executor's own runnables.  These runnables may never return until
+    // netty shuts down.  Any created continuations will be open until that time preventing traces
+    // from being reported
+    if (!task.getClass()
+        .getName()
+        .startsWith("io.netty.util.concurrent.SingleThreadEventExecutor$")) {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: I think this is executed in a very hot path... Should we be concerned about a string startswith call?

File: dd-java-agent/instrumentation/twitter-promise/twitter-promise.gradle [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+dependencies {
+  compileOnly group: 'com.twitter', name: 'util-core_2.11', version: '19.1.0'
+
+  testCompile project(':dd-java-agent:instrumentation:netty-4.1')
+  testCompile project(':dd-java-agent:instrumentation:java-concurrent')
+  testCompile project(':dd-java-agent:instrumentation:trace-annotation')
+
+  testCompile group: 'com.twitter', name: 'util-core_2.11', version: '19.1.0'
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/twitter-promise/twitter-promise.gradle]: Is it possible to write any useful tests for this project?

**Comment by randomanderson** [comment]
Replacing with finatra instrumentation

**Pull request closed without merging**